### PR TITLE
add log rotation to docker driver log defaults

### PR DESCRIPTION
**Pull request by jazzyfresh**
## Overview
* Nomad uses Docker log defaults if no user log configuration provided
* Docker log driver defaults to `json-file` with no log rotation, which will make logs grow infinitely
* **This adds log rotation to the default Docker logging**

## Behavior
* rotate docker logs with `max-file=2` and `max-size=2m` by default
* "by default" = only add log rotation if no user config provided (neither log Type nor log Config is set)

## Implementation
* https://docs.docker.com/config/containers/logging/configure/

### CODE REVIEW

File: drivers/docker/driver.go [hunk 1]
```diff
@@ -740,6 +740,20 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
 		Config: driverConfig.Logging.Config,
 	}
 
+	if hostConfig.LogConfig.Type == "json-file" || hostConfig.LogConfig.Type == "" {
```
[the remark this answers is not in the document]
Comment (jazzyfresh) [reply drivers/docker/driver.go]: i thought `json-file` is the default for docker https://docs.docker.com/config/containers/logging/json-file/

[the remark this answers is not in the document]
Comment (jazzyfresh) [reply drivers/docker/driver.go]: oh or if type is unset/unconfigured by the user, do we set it to `syslog` somewhere?

[the remark this answers is not in the document]
Comment (notnoop) [reply drivers/docker/driver.go]: Indeed, starting with Nomad 0.9.0, we use the docker daemon setting by default, which defaults to `json-file`.  However, users can configure that using daemon.json.

I'd suggest that we check daemon setting as well as the task logging setting.  You can inspect the daemon logging default from `dockerInfo` in fingerprint function[1] and store it in a field to inspect here.

Additionally, I'd suggest setting the values (e.g. `max-file`) only if not set by user.  This allows users to override our defaults if they so choose.

[1] https://github.com/hashicorp/nomad/blob/51d4fa0fc180d9a85fbee3978a2cea5265befb7d/drivers/docker/fingerprint.go#L149-L162

[the remark this answers is not in the document]
Comment (endocrimes) [reply drivers/docker/driver.go]: Oh cool - we should update the docs then, because we currently say syslog is the default.

File: drivers/docker/driver.go [hunk 2]
```diff
@@ -740,6 +740,20 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
...
+		logger.Debug("configuring docker json-file logs")
+		hostConfig.LogConfig.Type = "json-file"
+		if hostConfig.LogConfig.Config == nil {
+			hostConfig.LogConfig.Config = make(map[string]string)
+		}
+		hostConfig.LogConfig.Config["max-file"] = "3"
+		hostConfig.LogConfig.Config["max-size"] = "10m"
```
Comment (schmichael) [inline drivers/docker/driver.go]: Minimize these resources by default as they're in addition to the logs Nomad and only used as a buffer during agent restarts. (If the operator wants to use them directly they can manually configure them.)

Maybe:
```go
		hostConfig.LogConfig.Config["max-file"] = "2"
		hostConfig.LogConfig.Config["max-size"] = "2m"
```

Comment (jazzyfresh) [reply drivers/docker/driver.go]: fixed

File: drivers/docker/driver_test.go [hunk 3]
```diff
@@ -1014,25 +1014,44 @@ func TestDockerDriver_CreateContainerConfig_Logging(t *testing.T) {
 	cases := []struct {
 		name           string
 		loggingConfig  DockerLogging
-		expectedDriver string
+		expectedConfig DockerLogging
```
Comment (notnoop) [inline drivers/docker/driver_test.go]: Maybe add some tests for non-empty config and test merging logic if we want to override `Config != nil` as well?

File: drivers/docker/driver.go [hunk 4]
```diff
@@ -740,6 +740,14 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
 		Config: driverConfig.Logging.Config,
 	}
 
+	if hostConfig.LogConfig.Type == "" && hostConfig.LogConfig.Config == nil {
+		logger.Trace("no docker log driver provided, defaulting to json-file")
+		hostConfig.LogConfig.Type = "json-file"
+		hostConfig.LogConfig.Config = make(map[string]string)
+		hostConfig.LogConfig.Config["max-file"] = "2"
+		hostConfig.LogConfig.Config["max-size"] = "2m"
```
Comment (notnoop) [inline drivers/docker/driver.go]: You can also do the following that I find a bit more readable

```go
hostConfig.LogConfig.Config = map[string]string {
  "max-file": "2",
  "max-size": "2m"
}
```

Comment (jazzyfresh) [reply drivers/docker/driver.go]: oof, good catch
